src/utils.rs

```diff
@@ -130,6 +130,53 @@ impl<P: AsRef<Path> + std::fmt::Debug> ToUrl for P {
     }
 }
 
+fn newlines_in_range(range: Range) -> u64 {
+    range.end.line - range.start.line
+}
+
+fn character_difference_in_last_edit_line(e: &TextEdit) -> u64 {
+    (e.new_text.lines().last().unwrap_or("").chars().count() as u64)
+        - (e.range.end.character
+            + if e.range.start.line == e.range.end.line {
+                e.range.start.character
+            } else {
+                0
+            })
+}
+
+fn calc_cursor_column_after_TextEdit(edit: &TextEdit, prev_position: (u64, u64)) -> u64 {
+    // Edit happens in same line and before cursor
+    if (edit.range.end.line == prev_position.0) && (edit.range.end.character <= prev_position.1) {
+        prev_position.1 + character_difference_in_last_edit_line(edit)
+    } else {
+        prev_position.1
+    }
+}
+
+fn newlines_in_str(str: &str) -> u64 {
```

Comment:
I wonder if this could be replaced with just `some_str.matches('\n').count()`, thoughts?

Reply:
I thought max(\n,\r) would work for both Windows (\r\n), Unix (\n) and Apple (\r) encoding. I think the LSP protocol does not normalize the line endings. From the spec:

Text Documents

The current protocol is tailored for textual documents whose content can be represented as a string. There is currently no support for binary documents. A position inside a document (see Position definition below) is expressed as a zero-based line and character offset. The offsets are based on a UTF-16 string representation. So a string of the form a𐐀b the character offset of the character a is 0, the character offset of 𐐀 is 1 and the character offset of b is 3 since 𐐀 is represented using two code units in UTF-16. To ensure that both client and server split the string into the same line representation the protocol specifies the following end-of-line sequences: ‘\n’, ‘\r\n’ and ‘\r’.

Reply:
Sorry, what I meant to say was that `matches` looks simpler than iterating `chars`, may be slower though, but maybe it's still worth it.

Reply:
Sorry, I just realized this also. Will fix it! :+1: